ts-rs/src/export.rs

```diff
@@ -26,6 +29,51 @@ pub enum ExportError {
     ManifestDirNotSet,
 }
 
+/// Exports `T` to the file specified by the `#[ts(export_to = ..)]` attribute.
+/// Additionally, all dependencies of `T` will be exported as well.
+/// TODO: This might cause a race condition:
+///       If two types `A` and `B` are `#[ts(export)]` and depend on type `C`,
+///       then both tests for exporting `A` and `B` will try to write `C` to `C.ts`.
+///       Since rust, by default, executes tests in paralell, this might cause `C.ts` to be corrupted.
+pub(crate) fn export_type_with_dependencies<T: TS + ?Sized + 'static>() -> Result<(), ExportError> {
+    // exports T, then recursively calls itself with all of its dependencies
+    fn export_recursive<T: TS + ?Sized + 'static>(
+        seen: &mut HashSet<TypeId>,
+    ) -> Result<(), ExportError> {
+        struct Visit<'a> {
+            seen: &'a mut HashSet<TypeId>,
+            error: Option<ExportError>,
+        }
+        impl<'a> TypeVisitor for Visit<'a> {
+            fn visit<T: TS + 'static + ?Sized>(&mut self) {
+                // if an error occurred previously, or the type cannot be exported (its a primitive),
+                // we return
+                if self.error.is_some() || T::EXPORT_TO.is_none() {
+                    return;
+                }
+                self.error = export_recursive::<T>(self.seen).err();
+            }
+        }
+
+        if seen.insert(TypeId::of::<T>()) {
+            export_type::<T>()?;
+            let mut visitor = Visit { seen, error: None };
+            T::dependency_types().for_each(&mut visitor);
+
+            if let Some(e) = visitor.error {
+                Err(e)
+            } else {
+                Ok(())
+            }
+        } else {
+            Ok(())
+        }
+    }
```

Comment:
Are these definitions required to be nested?

Reply:
No, not at all. I just wanted to encapsulate `Visit` and the `TypeVisitor` impl for it, since it has no uses outside of that function. Happy to do it differently if you dislike it.

Reply:
> No, not at all. I just wanted to encapsulate Visit and the TypeVisitor impl for it, since it has no uses outside of that function.

I see. Maybe a `mod` would help with that?

Reply:
> I see. Maybe a mod would help with that?

I did it like this, what do you think?

```rs
pub(crate) use recursive_export::export_type_with_dependencies;
mod recursive_export {
    use std::{any::TypeId, collections::HashSet};

    use super::export_type;
    use crate::{
        typelist::{TypeList, TypeVisitor},
        ExportError, TS,
    };

    struct Visit<'a> {
        seen: &'a mut HashSet<TypeId>,
        error: Option<ExportError>,
    }

    impl<'a> TypeVisitor for Visit<'a> {
        fn visit<T: TS + 'static + ?Sized>(&mut self) {
            // if an error occurred previously, or the type cannot be exported (it's a primitive),
            // we return
            if self.error.is_some() || T::EXPORT_TO.is_none() {
                return;
            }

            self.error = export_recursive::<T>(self.seen).err();
        }
    }

    /// Exports `T` to the file specified by the `#[ts(export_to = ..)]` attribute.
    /// Additionally, all dependencies of `T` will be exported as well.
    ///
    /// TODO: This might cause a race condition:
    ///       If two types `A` and `B` are `#[ts(export)]` and depend on type `C`,
    ///       then both tests for exporting `A` and `B` will try to write `C` to `C.ts`.
    ///       Since rust, by default, executes tests in paralell, this might cause `C.ts` to be corrupted.
    pub(crate) fn export_type_with_dependencies<T: TS + ?Sized + 'static>(
    ) -> Result<(), ExportError> {
        let mut seen = HashSet::new();
        export_recursive::<T>(&mut seen)
    }

    // exports T, then recursively calls itself with all of its dependencies
    fn export_recursive<T: TS + ?Sized + 'static>(
        seen: &mut HashSet<TypeId>,
    ) -> Result<(), ExportError> {
        if !seen.insert(TypeId::of::<T>()) {
            return Ok(());
        }

        export_type::<T>()?;

        let mut visitor = Visit { seen, error: None };
        T::dependency_types().for_each(&mut visitor);

        if let Some(e) = visitor.error {
            Err(e)
        } else {
            Ok(())
        }
    }
}
```

Reply:
100% fine with me! You can commit that, or I can copy it over, whatever is easier for you.